zip-0221.rst

```diff
@@ -0,0 +1,585 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This ZIP introduces a new header version. It is identical to the current v4 header [#zcashBlock]_, except for the following changes:
+
+1. The version number is changed to `5`.
+2. ``hashSaplingFinalRoot`` is replaced by ``hashChainHistoryRoot``, as described above.
+
+The new block header format is:
+
+.. code-block:: C
+
+    class CBlockHeader
+    {
+    public:
+        // header
+        static const size_t HEADER_SIZE=4+32+32+32+4+4+32; // excluding Equihash solution
+        static const int32_t CURRENT_VERSION=5;
+        int32_t nVersion;
+        uint256 hashPrevBlock;
+        uint256 hashMerkleRoot;
+        uint256 hashChainHistoryRoot;
+        uint32_t nTime;
+        uint32_t nBits;
+        uint256 nNonce;
+        std::vector<unsigned char> nSolution;
+        ...
+    }
+
+
+Rationale
+-----------
+
+Tree nodes
+~~~~~~~~~~~
+
+Nodes in the commitment tree are canonical and immutable. They are cheap to generate, as (with the exception of ``nShieldedTxCount``) all metadata is already generated during block construction and/or checked during block validation. Nodes are relatively compact in memory. Approximately 140,000 blocks have elapsed since Sapling activation. Assuming a 164 byte commitment to each of these, we would have generated approximately 24 MB of additional storage cost for the set of leaf nodes (and an additional ~24 MB for storage of intermediate nodes).
+
+``hashSubtreeCommitment`` forms the strucuture of the commitment tree. Other metadata commitments were chosen to serve specific purposes. Variable-length commitments are placed last, so that most metadata in a node can be directly indexed. We considered using fixed-length commitments here, but opted for variable-length, in order to marginally reduce the memory requirements for managing and updating the commitment trees.
+
+In leaf nodes, some information is repeated. We chose to do this so that leaf nodes could be treated identically to internal and root nodes for all algorithms and (de)serializers. Leaf nodes are easily identifiable, as they will show proof of work in the ``hashSubtreeCommitment`` field, and their block range (calculated as ``nLatestHeight - nEarliestHeight + 1``) will be precisely 1. For the same reason, we change the semantics of ``hashSubtreeCommitment`` in leaf nodes to commit
+
+Personalized BLAKE2b-256 was selected to match existing ZCash conventions. Adding the consensus branch ID to the hash personalization string ensures that valid nodes from one branch cannot be used to make false statements about parallel consensus branches.
+
+FlyClient Requirements and Recommendations
+===============================================
+These commitments enable FlyClient in the variable-difficulty model. Specifically, they allow light clients to reason about application of the difficulty adjustment algorithm over a range of blocks. They were chosen via discussion with an author of the FlyClient paper.
```

Comment:
Hi, IMO adding these difficulty adjustment fields on the node brings no more security than the `Bits` field from the header.

If an attacker tries to forge a series of blocks with valid but lower pow, he can also forge those difficulty adjustment variables to make the difficulty target lower.

Please correct me if I'm wrong.

Reply:
Hey, I'd recommend reading the FlyClient paper if you haven't already. FlyClient verifiers get confidence in the correctness of a history commitment by requiring the prover to open a number of commitments from the tree. If the prover has forged the commitment, they will be able to show valid leaves with negligible probability. 

FlyRelays, on the other hand, are forced to use an optimistic verification game, due to the on-chain expense of opening commitments. Either way we get confidence in the correctness of the tree.

Once we're sure the tree is correct, then a series of blocks with lower PoW but valid difficulty adjustments is not a forgery, it's just a minority fork which we can easily ignore.

Reply:
Thanks for the reply, I had read the FlyClient paper.

In the original paper, the MMR is encoding with a `total difficulty` variable that indicates the total difficulty under a node. But this ZIP uses difficulty adjustments variables instead of the `total difficulty`, so we can calculate the difficulty.

My question is what the advantage it brings is?

In my understanding, this design intends to let the verifier calculate the difficulty itself, to avoid to be fooled by the prover with a lower difficulty. (Is this right? I can't find details explain from this ZIP). But the difficulty adjustment variables are also from the prover side, so verifier can't trust these variables on the first until the verifier randomly verify the chain.

So I' am confused, add more fields in merkle tree only increases the merkle proof size, but no help on the random verification of chain, I can't see the advantage to do this. 

Maybe I misunderstand the design, expect an explanation : )